### PR TITLE
Fix meson location detection from other meson tools

### PR DESCRIPTION
**Pull request by alyst**
ArchLinux meson package installs `wraptool` to `/usr/bin`, so when I try to run it:
```
> wraptool list
Traceback (most recent call last):
  File "/usr/bin/wraptool", line 17, in <module>
    from mesonbuild.wrap import wraptool
  File "/usr/lib/python3.6/site-packages/mesonbuild/wrap/wraptool.py", line 22, in <module>
    from .wrap import API_ROOT, open_wrapdburl
  File "/usr/lib/python3.6/site-packages/mesonbuild/wrap/wrap.py", line 22, in <module>
    from ..mesonlib import Popen_safe
  File "/usr/lib/python3.6/site-packages/mesonbuild/mesonlib.py", line 61, in <module>
    meson_command = python_command + [detect_meson_py_location()]
  File "/usr/lib/python3.6/site-packages/mesonbuild/mesonlib.py", line 28, in detect_meson_py_location
    raise RuntimeError(c_fname)
```

That's because `detect_meson_py_location()` assumes that the only script that could be run like that is `meson` or `meson.py`.
The PR changes the logic of this function a bit:
 * first, it tries to get the absolute path to the meson tool folder
 * then it tries to find meson executable in that folder

### CODE REVIEW

File: mesonbuild/mesonlib.py [hunk 1]
```diff
@@ -25,25 +25,33 @@
...
-            if '.' in os.environ['PATH'].split(':'):
-                p, f = os.path.split(in_path_exe)
-                if p == '' or p == '.':
-                    return os.path.join(os.getcwd(), f)
-            return in_path_exe
-        # $ python3 ./meson.py <args>
-        if os.path.exists(c):
-            return os.path.join(os.getcwd(), c)
+            if ('.' in os.environ['PATH'].split(':') and
+                (m_dir == '' or m_dir == '.')):
```
Comment (ignatenkobrain) [inline mesonbuild/mesonlib.py]: __[Flake8]__

[E129] visually indented line with same indent as next logical line

File: mesonbuild/mesonlib.py [hunk 2]
```diff
@@ -25,25 +25,33 @@
...
+                m_dir = os.getcwd()
+    else:
+        m_dir = os.path.abspath(c_dir)
+
+    # find meson in m_dir
+    if not m_dir is None:
```
Comment (ignatenkobrain) [inline mesonbuild/mesonlib.py]: __[Flake8]__

[E714] test for object identity should be 'is not'

**Comment by alyst** [comment]
Fixes #2810

**Comment by nirbheek** [comment]
This would fix https://github.com/mesonbuild/meson/issues/2810 and https://github.com/mesonbuild/meson/issues/2761.

**Comment by jpakkane** [comment]
Test failure seems very relevant.

**Comment by alyst** [comment]
Yes, I'm just trying to get windows system to debug it.

**Comment by jpakkane** [comment]
[These are what I use.](https://developer.microsoft.com/en-us/microsoft-edge/tools/vms/)

File: mesonbuild/mesonlib.py [hunk 3]
```diff
@@ -17,37 +17,52 @@
...
-        if os.path.isabs(c):
-            return c
-        # $ meson <args> (gets run from /usr/bin/meson)
+    c_dir, c_fname = os.path.split(c)
+    if c_dir == '' or c_dir == '.':
+        # if python expects posix path, but windows path given
+        nt_dir, nt_fname = ntpath.split(c)
+        if nt_dir != '' and nt_dir != '.':
+            c_dir, c_fname = nt_dir, nt_fname
+
```
Comment (ignatenkobrain) [inline mesonbuild/mesonlib.py]: __[Flake8]__

[W293] blank line contains whitespace

**Comment by alyst** [comment]
Now it should be fixed.

The current logic:
 1. get the path to the folder with the current executable (might be `meson`, `wraptool`, legacy scripts (`mesontest`), test scripts (`run_unittests.py`))
 2. if the path is not empty but not absolute, normalize it (i.e. expand `.` to cwd)
 3. if the path is empty, use `shutil.which(script)` to get the path (`m_dir`). Previously, if `m_dir == '.'`, it was expanded to cwd only when `PATH` contained `'.'`. Now this check is dropped (we trust `shutil.which`). Otherwise on windows when the executable is `run_unittests.py` the detected meson location was `./meson.py`.
 4. If `m_dir` does not have `meson[.py]`, use `<path to mesonlib.py>/../meson.py`

There is one more thing that I think is worth doing: `run_tests.py` should invoke `run_unitests.py` and other scripts as `os.path.join('.', 'run_unittests.py')`.
With this change we would ensure that tests always call `meson.py` from the same folder as the test scripts and avoid relying on `shutich.which('run_unittests.py')` result.

**Pull request merged**